### PR TITLE
[batch] Update flock strategy

### PR DESCRIPTION
**Pull request by chrisvittal**
flock(2) does not prevent other processes from doing anything to the
locked file unless they also gate usages on an flock call. Update Flock
to open only one descriptor, the directory that is the parent of `path`,
creating said directory if it doesn't exist.

The semantics have changed slightly. Previously if path was an already
extant directory, it would be opened and that descriptor would be
flocked. Now, the parent is always locked.

### CODE REVIEW

File: batch/batch/worker/flock.py [hunk 1]
```diff
@@ -3,40 +3,29 @@
 import argparse
 import subprocess as sp
 
+from pathlib import Path
```
Comment (chrisvittal) [inline batch/batch/worker/flock.py:6]: The path manipulations are simpler in this version than they were in my first draft. I'm not sure this is needed, and `os.path` may be fine.

**Comment by jigold** [comment]
I can't remember why I did it the other way, but your way seems simpler.

**Review by jigold — DISMISSED** [review]

**Comment by jigold** [comment]
Oh. I remember why now. When we were using rsync I was worried someone would simultaneous try to do this:

cp gs://bucket/a/b/
cp gs://bucket/a/

Then they would clobber each other.

**Comment by jigold** [comment]
But if we're copying by file, then maybe it's fine to use your strategy. What do you think?

**Comment by chrisvittal** [comment]
> Oh. I remember why now. When we were using rsync I was worried someone would simultaneous try to do this:
> 
> cp gs://bucket/a/b/
> cp gs://bucket/a/
> 
> Then they would clobber each other.

Can you flesh this out, I'm not sure I understand the sequence of commands here.

**Comment by jigold** [comment]
Let's say we have the following structure

gs://bucket/a/b/foo.txt
gs://bucket/a/b/bar.txt
gs://bucket/a/baz.txt


In the old design with rsync, we could be doing in parallel the following:

```
gsutil rsync gs://bucket/a
```

and

```
gsutil rsync gs://bucket/a/b/
```

I was worried we would delete files in the rsync in the middle of one job doing the copying. Basically some kind of race condition.

However, with the new design of the cache, this might not be a problem where we're not using gsutil rsync but instead writing our own version of rsync. Cotton's fleshing out the design of that.

**Comment by chrisvittal** [comment]
I believe that this patch and the original version both prevent this. They would both lock the parent directory of `path` and thus would prevent concurrent copying.

Also, it appears that as long as we don't use the `-d` option, and the filenames are unique, `gsutil rsync` kinda already does what we want.

I also feel like the orignal approach was way too aggressive, it seems like it was serializing _all_ locked filesystem operations in the entire filesystem subtree since it would wait for the lock on every parent other than `/` of the requested file.

**Review by jigold — APPROVED** [review]

**Pull request merged**